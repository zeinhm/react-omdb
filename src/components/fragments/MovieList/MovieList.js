import React, { useState } from 'react'
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom'
import Card from '../../elements/Card'
import Modal from '../../elements/Modal'
import defaultImg from '../../../assets/default-poster.jpg'
import './styles.scoped.css'

export default function MovieList(props) {
  const history = useHistory()
  const { dataShownMovies } = props
  const [openModal, setOpenModal] = useState(false)
  const [imgSrc, setImgSrc] = useState('')
  const defaultPoster = (e) => {
    e.target.src = defaultImg
  }
  const onClick = id => {
    history.push(`detail/?id=${id}`)
  }
  const onClickImg = imgSrc => {
    setImgSrc(imgSrc)
    setOpenModal(true)
  }
  const onCloseModal = () => {
    setOpenModal(false)
  }
  return (
    <div>
      <ul>
        {
          dataShownMovies.length >= 1 && dataShownMovies.map((data, key) => (
            <li key={key}>
              <Card className='card-movie'>
                <figure>
                  <img
                    alt="Poster"
                    onClick={()=> onClickImg(data.Poster)}
                    onError={defaultPoster}
                    src={data.Poster}
                    />
                  <figcaption>{data.Title}</figcaption>
                  <button onClick={() => onClick(data.imdbID)}>Detail</button>
                </figure>
              </Card>
            </li>
          )) 
        }
      </ul>
      <ModalPoster
        defaultPoster={defaultPoster}
        imgSrc={imgSrc}
        onClose={()=>onCloseModal(false)}
        open={openModal}
        />
    </div>
  )
}

MovieList.defaultProps = {
  dataShownMovies: []
}

MovieList.propTypes = {
  dataShownMovies: PropTypes.array
}

export function ModalPoster({ defaultPoster, open, onClose, imgSrc }) {
  return (
    <Modal className='modal-poster' open={open}>
      <img
        alt="Poster"
        onError={defaultPoster}
        src={imgSrc}
        />
      <button onClick={() => onClose()}>Close</button>
    </Modal>
  );
}

ModalPoster.defaultProps = {
  defaultPoster: () => { },
  onClose: () => { },
  open: false,
};

ModalPoster.propTypes = {
  defaultPoster: PropTypes.func,
  onClose: PropTypes.func,
  open: PropTypes.bool,
};