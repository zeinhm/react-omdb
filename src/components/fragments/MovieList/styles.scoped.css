.card-movie {
  background-color: #FFFFFF;
  border-radius: 20px 0;
  box-shadow: 0.6875rem 0.5625rem 1.3125rem rgba(173, 182, 217, 0.5);
  margin: 2rem 0;
  padding: 0 0 1rem;
  width: 15.625rem;
}

.card-movie figure {
  border-radius: 15px 0;
}

.card-movie figure img {
  border-radius: 15px 0 0;
  cursor: pointer;
  width: 100%;
}

.card-movie figcaption {
  margin: 1rem;
  text-align: center;
}

.card-movie figure > button {
  background: linear-gradient(to right top, #d16ba5, #c777b9, #ba83ca, #aa8fd8, #9a9ae1, #8aa7ec, #79b3f4, #69bff8, #52cffe, #41dfff, #46eefa, #5ffbf1);
  border: none;
  border-radius: 10px 0;
  color: #fff;
  display: block;
  height: 2.5rem;
  margin: 0 5%;
  padding: 0.4rem;
  text-align: center;
  width: 90%;
}

.card-movie figure > button:hover {
  background: linear-gradient(to bottom right, #d16ba5, #c777b9, #ba83ca, #aa8fd8, #9a9ae1, #8aa7ec, #79b3f4, #69bff8, #52cffe, #41dfff, #46eefa, #5ffbf1);
  color: #fff;
  cursor: pointer;
  font-weight: bold;
  text-decoration: none;
}

.modal-poster {
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-bottom: 5rem;
  padding: 0;
  text-align: center;
  width: fit-content;
}

.modal-poster img {
  width: 25rem;
}

.modal-poster button {
  height: 3rem;
}

.modal-poster button:hover {
  cursor: pointer;
}

@media(max-width: 424px) {
  .modal-poster img {
    width: 18.5rem;
  }
}