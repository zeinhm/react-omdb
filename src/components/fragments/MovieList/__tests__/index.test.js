import React from 'react';
import { useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import MovieList, { ModalPoster } from '../MovieList';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(cb => {
    cb({});
    return {
      state: {
        dataMovieDetail: {},
        isLoadingMovieDetail: false
      }
    };
  }),
}));

describe('src/components/fragments/MovieList', ()=> {
  test('MovieList', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<MovieList />);
    
    useSelector.mockImplementationOnce(() => ({
      state: {
        dataMovieDetail: {},
        isLoadingMovieDetail: false
      }}));
    expect(tree).toMatchSnapshot();
  });
  test('ModalPoster', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<ModalPoster />);
    ModalPoster.defaultProps.onClose();
    expect(tree).toMatchSnapshot();

    const result = ModalPoster({ ...ModalPoster.defaultProps, open: true });
    result.props.onClose();
    result.props.children[1].props.onClick();
  });
});
