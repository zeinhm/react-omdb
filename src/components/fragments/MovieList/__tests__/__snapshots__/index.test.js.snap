// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fragments/MovieList ModalPoster 1`] = `
<Modal
  className="modal-poster"
  onClose={[Function]}
  open={false}
>
  <img
    alt="Poster"
    onError={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Close
  </button>
</Modal>
`;

exports[`src/components/fragments/MovieList MovieList 1`] = `
<div>
  <ul />
  <ModalPoster
    defaultPoster={[Function]}
    imgSrc=""
    onClose={[Function]}
    open={false}
  />
</div>
`;
